File: TheDoomedKingdom/Classes/Models/SmartJSONAdapter.m
//
//  SmartJSONAdapter.m
//  TheDoomedKingdom
//
//  Created by Pavel Stoma on 1/22/18.
//  Copyright © 2018 PP. All rights reserved.
//

#import "SmartJSONAdapter.h"

@implementation SmartJSONAdapter

- (NSSet *)serializablePropertyKeys:(NSSet *)propertyKeys
                           forModel:(id<SmartJSONAdapting>)model
{
    if ([[model class] respondsToSelector:@selector(withoutNil)]
        && [[model class] withoutNil]) {
        NSMutableSet *propertyKeys_ = [propertyKeys mutableCopy];
        
        [propertyKeys enumerateObjectsUsingBlock:^(NSString *_Nonnull propertyKey,
                                                   BOOL * _Nonnull stop) {
            if ([[NSNull null] isEqual:model.dictionaryValue[propertyKey]]) {
                [propertyKeys_ removeObject:propertyKey];
            }
        }];
        propertyKeys = [propertyKeys_ copy];
    }
    if ([[model class] respondsToSelector:@selector(propertyKeysForJSONRepresentation)]) {
        NSMutableSet *propertyKeys_ = [propertyKeys mutableCopy];
        [propertyKeys_ minusSet:[[model class] propertyKeysForJSONRepresentation]];
        
        propertyKeys = [propertyKeys_ copy];
    }
    return propertyKeys;
}

@end

File: Calamity/Managers/GoogleDocs/Services/SmartJSONAdapter.swift
//
//  SmartJSONAdapter.swift
//  Calamity
//
//  Created by Pavel Stoma on 3/12/19.
//  Copyright © 2019 GameJam. All rights reserved.
//

import Mantle

protocol SmartJSONAdapting: MTLJSONSerializing {
    static func withoutNil() -> Bool
}

class SmartJSONAdapter: MTLJSONAdapter {
    func serializable(propertyKeys: Set<String>, forModel model: GoogleBaseModel) -> Set<String> {
        guard type(of: model).withoutNil() else {
            return propertyKeys
        }
        
        return propertyKeys.filter { (key) -> Bool in
            guard let value = model.dictionaryValue[key] else { return false }
            return !(value is NSNull)
        }
    }
}
